File: 020_valid_parentheses.py

# define parentheses object and stack array 
# loop s, if in object-> means it's open bracket, push closing in to stack 
# else if its closing bracket, it has to equal the top of stack (make sure stack is not empty before pop) , else return false
# return true if stack empty at end of loop


# time: O(n)
# space: O(n)

class Solution:
    def isValid(self, s: str) -> bool:
        p_map = {
            "(" : ")",
            "[" : "]",
            "{" : "}",
        }

        stack = [] 

        for char in s:
            if char in p_map:
                stack.append(p_map[char])
            else:
                if not stack or char != stack.pop():
                    return False 
        
        return not stack

File: 020_validParentheses.js
// # define parentheses object and stack array 
// # loop s, if in object-> means it's open bracket, push closing in to stack 
// # else if its closing bracket, it has to equal the top of stack (make sure stack is not empty before pop) , else return false
// # return true if stack empty at end of loop


// # time: O(n)
// # space: O(n)

var isValid = function(s) {
    const obj = {
        "{": "}",
        "(": ")",
        "[": "]"
    }

    const stack = [] 

    for(let i = 0; i < s.length; i++) {
        if (obj.hasOwnProperty(s[i])) {
            stack.push(obj[s[i]])
        } else {
            if (stack.length === 0 || s[i] !== stack.pop()) {
                return false
            }
        }
    }
    return stack.length === 0
};